### PR TITLE
Fix seg fault houghlinespointset

### LINKED ISSUE #21029: Segmentation Fault with `HoughLinesPointSet`
##### System information (version)

- OpenCV => 4.5
- Operating System / Platform => Linux 64 Bit
- Compiler => ? (whatever was used to build the opencv-python package)

##### Detailed description

The function `HoughLinesPointSet` crashes randomly with Segmentation Faults.

The Hough accumulator (histogram of angles and distances that define lines) is populated using a pointer and an index calculated from `r`, the distance to a candidate line. The accumulator is initialized independently from the point coordinates, however `r` depends on the point coordinates, and thus can lead to memory access outside the accumulator.

https://github.com/opencv/opencv/blob/6f4160c0140c8ef88d777ba2ee5a605bf345de64/modules/imgproc/src/hough.cpp#L976

##### Steps to reproduce

```python
import sys
import numpy as np
import cv2
print(sys.version)
print(np.__version__)
print(cv2.__version__)

from cv2 import HoughLinesPointSet

def foo():
    points = np.array([(100, 100), (1000, 1000), (10000, 10000), (100000, 100000)])

    # Create array shape required by opencv
    # https://stackoverflow.com/questions/47402445/need-help-in-understanding-error-for-cv2-undistortpoints/47403282#47403282
    points = points[:, np.newaxis, :].astype('float32')

    ## Hough transform config
    # rho: distance origin to closest point on line
    rhoMin = 0
    rhoMax = 10
    rhoStep = 0.1

    # theta: angle between x-axis and line connection closest point to origin
    # limit line search to horizontal lines
    thetaMin = np.deg2rad(85)
    thetaMax = np.deg2rad(95)
    thetaStep = np.deg2rad(1)

    lines_max = 5
    threshold = 100

    HoughLinesPointSet(points, lines_max, threshold, rhoMin, rhoMax, rhoStep, thetaMin, thetaMax, thetaStep)


for i in range(1000):
    foo()
```
Output:
```
3.9.6 (default, Jul 30 2021, 16:35:19) 
[GCC 7.5.0]
1.21.1
4.5.4-dev
Segmentation fault (core dumped)
```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by tv3141**
### Pull Request Readiness Checklist

Closes #21029 

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/imgproc/src/hough.cpp [hunk 1]
```diff
@@ -975,7 +975,9 @@ void HoughLinesPointSet( InputArray _point, OutputArray _lines, int lines_max, i
         for(int n = 0; n < numangle; n++ )
         {
             int r = cvRound( point.at(i).x  * tabCos[n] + point.at(i).y * tabSin[n] - irho_min);
-            accum[(n+1) * (numrho+2) + r+1]++;
+            if ( r >= 0 && r <= numrho) {
```
Comment (alalek) [inline modules/imgproc/src/hough.cpp:978]: Could you please to put regression test for this case here: https://github.com/opencv/opencv/blob/3.4.16/modules/imgproc/test/test_houghlines.cpp#L301

```
TEST(HoughLinesPointSet, regression_21029)
{
    std::vector<Point2f> points;
    points.push_back(Point2f(100, 100));
    points.push_back(Point2f(1000, 1000));
    points.push_back(Point2f(10000, 10000));
    points.push_back(Point2f(100000, 100000));

    double rhoMin = 0;
    double rhoMax = 10;
    double rhoStep = 0.1;

    double thetaMin = 85 * CV_PI / 180.0;
    double thetaMax = 95 * CV_PI / 180.0;
    double thetaStep = 1 * CV_PI / 180.0;

    int lines_max = 5;
    int threshold = 100;

    Mat lines;

    HoughLinesPointSet(points, lines,
        lines_max, threshold,
        rhoMin, rhoMax, rhoStep,
        thetaMin, thetaMax, thetaStep
    );

    EXPECT_FALSE(lines.empty());
}
```

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

As a bugfix this patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into 4.x regularly (weekly/bi-weekly).

Please:
- change "base" branch of this PR: 4.x => 3.4 (use "Edit" button near PR title)
- rebase your commits from 4.x onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/4.x`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

File: modules/imgproc/src/hough.cpp (hunk 1 above)
Comment (tv3141) [reply modules/imgproc/src/hough.cpp:978]: I added the test but changed `EXPECT_FALSE(lines.empty());` to `EXPECT_TRUE(lines.empty());` as no lines will be found.

**Comment by alalek** [comment]
Thank you for update!

Please take a look on doxygen warnings from "Docs" builder: http://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/32796

**Review by alalek — APPROVED** [review]
Thank you!

**Comment by tv3141** [comment]
@alalek Thanks for walking me through the process.

**Pull request merged**